### ISSUE TITLE
Document the return value of exported function as a separate entity

### ISSUE THREAD

**Issue by kasperlewau** [esdoc/esdoc#296]
*This is probably not an issue and would probably make a better fit at a place like StackOverflow, but there doesn't seem to be a lot of traction on there for ESDoc - what with the lack of a suitable tag and all.* 

--- 

I'm currently in the process of rewriting an Angular 1.x codebase into ES2015. We've identified two core concepts that fit the `class` bill incredibly well. We've also managed to identify a fairly decent way of hooking said class into the `service` layer intricacies that is Angular 1.x. 

In essence, this is what the code (read: boilerplate) looks like: 

```js
// FooModel.es6

/**
 * angular service wrapper fn. 
 * @ignore (wont work)
 */
export default function ($q, $x, $z, ...otherAngularDeps) {
  /**
   * Foo class.
   */
  class Foo {
    /**
     * Foo constructor fn
     * @param {object} props
     */
    constructor (props) {}
  }

  return Foo;
}
```
```js
// importer.es6
import Foo from './FooModel.es6';
angular.service('Foo', Foo);
```

Now, in a perfect setting I would be able to _ignore_ the outer wrapper function and only document the `Foo` class - this does not seem to work with `@ignore`, which makes total sense imho. 

Barring `@ignore` - any documentation I have for the `Foo` class gets lost whenever the wrapping function is introduced. If I break out the class definition from the wrapping functions scope, ESDoc picks it up just fine 🚀. 

Now, I _could_ ditch the wrapper function and slap the Angular dependencies into the constructor function of `Foo` (this is the 'generally accepted' approach of using ES2015 classes in angular@1.x), but that would introduce _a ton_ of boilerplate and exposed services through the instances of `Foo`. **Yuck!**

--- 

**tl/dr**: I'm looking for suggestions on how to document _the return value of a function as a separate entity_, yet keep said entity in the given function scope. 

*(Feel free to close this and send my sorry bum over to StackOverflow, this is borderline issue/non-issue as far as I'm concerned. Sorry bout that.)*

**Comment by typhonrt** [comment]
A little nit here `class Foo () {` is invalid `class Foo {` is what it should be... I suppose this doesn't change the documentation situation though. 

You just might be able to fudge the AST generated with a custom ESDoc plugin that modifies the AST in `onHandleAST`, but that is pretty funky in general. 

IE dynamically move the AST for the Foo ClassDeclaration seen here: https://astexplorer.net/#/DnVgOekcoZ

to this:
https://astexplorer.net/#/DnVgOekcoZ/1

You can see the generated AST on the right. What I'm suggesting though again a bit freaky is moving the ClassDeclaration to the body of the `Program` node before `ExportDefaultDeclaration`.

**Comment by kasperlewau** [comment]
> A little nit here class Foo () { is invalid class Foo {

Sorry bout that one. Edited the main post. Thanks!

I'm absolutely fine with moving around the AST - the freakiness doesn't bother me _that_ much as I'm not concerned with the intricacies of how angular injectables so the wrapping function goes in the bin as far as ESDoc goes. 

I'll have a look, see if there isn't already a plugin written for this type of problem. Barring that, let's see how far I get with wiring up a plugin. 

Cheers!

**Comment by typhonrt** [comment]
There won't be an existing ESDoc plugin to handle this situation to my knowledge. 

Perhaps an edge tip... ESDoc might officially upgrade to Babylon I hope soon. I have a [maintained fork](https://github.com/typhonjs-node-esdoc/esdoc) of ESDoc that uses Babylon 6 for the parser. 

To make the AST parsing easier and more reliable if using the Babylon fork for this particular case you could use a decorator:
https://astexplorer.net/#/DnVgOekcoZ/2

And that way you can parse the ClassDeclaration looking for the declarator tag.

**Comment by kasperlewau** [comment]
Alrighty, I think this can be closed right about now. Made something incredible primitive but it appears to get the job done of lifting out the class declaration, generating some docs for it as well as ignoring the wrapper function. Happy days! 🚀 

```js
/**
 * @ignore
 * @ngClassLift
 */ 
export default function FooWrapper () {
  class Foo {
  }
  return Foo;
}
```

Thanks @typhonrt for nudging me in the right direction! 

[plugin as a gist](https://gist.github.com/kasperlewau/569db8c3ab1dd358318bf90fc98fc75c)

**Comment by typhonrt** [comment]
@kasperlewau Nice... Get 'er done! :8ball:

**Comment by mhelvens** [comment]
I'm having the same issue. I'll see if I can adapt @kasperlewau's plugin for now (thanks!), but I'm a bit confused about the basic problem.

Even if I don't use `@ignore` on the wrapper function, the class inside isn't documented. Doesn't ESDoc traverse inside of function bodies? Actually, I know it does, for some things. But I guess everything that's deemed 'local' is ignored.

Perhaps the JSDoc [`@global` tag](http://usejsdoc.org/tags-global.html) can be implemented?
